File: google-cast/src/main/java/com/theoplayer/sample/playback/cast/PlayerActivity.java
package com.theoplayer.sample.playback.cast;

import android.os.Bundle;
import android.util.Log;
import android.view.Menu;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;

import com.google.android.gms.cast.framework.CastButtonFactory;
import com.theoplayer.android.api.THEOplayerConfig;
import com.theoplayer.android.api.THEOplayerGlobal;
import com.theoplayer.android.api.cast.CastConfiguration;
import com.theoplayer.android.api.cast.CastIntegration;
import com.theoplayer.android.api.cast.CastIntegrationFactory;
import com.theoplayer.android.api.cast.CastStrategy;
import com.theoplayer.android.api.cast.chromecast.Chromecast;
import com.theoplayer.android.api.cast.chromecast.ChromecastConnectionCallback;
import com.theoplayer.android.api.event.chromecast.ChromecastEventTypes;
import com.theoplayer.android.api.event.player.PlayerEventTypes;
import com.theoplayer.android.api.player.Player;
import com.theoplayer.android.api.source.SourceDescription;
import com.theoplayer.android.api.util.Logger;
import com.theoplayer.sample.common.SourceManager;
import com.theoplayer.sample.playback.cast.databinding.ActivityPlayerBinding;

public class PlayerActivity extends AppCompatActivity {

    private static final String TAG = PlayerActivity.class.getSimpleName();

    private ActivityPlayerBinding viewBinding;
    private Player theoPlayer;
    private Chromecast theoChromecast;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        setTheme(com.theoplayer.sample.common.R.style.TheoTheme_Base);
        super.onCreate(savedInstanceState);

        // Inflating view and obtaining an instance of the binding class.
        viewBinding = DataBindingUtil.setContentView(this, R.layout.activity_player);

        // Gathering THEO objects references.
        theoPlayer = viewBinding.theoPlayerView.getPlayer();

        // Keep device screen on.
        viewBinding.theoPlayerView.setKeepScreenOn(true);

        // Enable all debug logs from THEOplayer.
        Logger theoDebugLogger = THEOplayerGlobal.getSharedInstance(this).getLogger();
        theoDebugLogger.enableAllTags();

        // THEOplayer automatically adds all available integrations to the player via the autoIntegrations() configuration. Here, we add the cast integration manually to configure the cast strategy.
        CastConfiguration configuration = new CastConfiguration.Builder().castStrategy(CastStrategy.AUTO).build();
        CastIntegration castIntegration = CastIntegrationFactory.createCastIntegration(viewBinding.theoPlayerView, configuration);
        theoPlayer.addIntegration(castIntegration);

        viewBinding.theoPlayerView.getCast();
        theoChromecast = viewBinding.theoPlayerView.getCast().getChromecast();

        // Configuring action bar.
        setSupportActionBar(viewBinding.toolbarLayout.toolbar);

        // Configuring THEOplayer playback with default parameters.
        configureTHEOplayer();
//        configureChromecast();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        super.onCreateOptionsMenu(menu);
        getMenuInflater().inflate(R.menu.activity_player_menu, menu);
        CastButtonFactory.setUpMediaRouteButton(getApplicationContext(), menu, R.id.castMenuItem);
        return true;
    }

    private void configureTHEOplayer() {
        // Coupling the orientation of the device with the fullscreen state.
        // The player will go fullscreen when the device is rotated to landscape
        // and will also exit fullscreen when the device is rotated back to portrait.
        viewBinding.theoPlayerView.getFullScreenManager().setFullScreenOrientationCoupled(true);

        // Allow background playback on the player to prevent Chromecast receiver from pausing when the app is backgrounded.
        viewBinding.theoPlayerView.getSettings().setAllowBackgroundPlayback(true);

        theoPlayer.setAutoplay(true);

        // Configuring THEOplayer with defined SourceDescription object.
        theoPlayer.setSource(SourceManager.Companion.getBIG_BUCK_BUNNY_HLS_WITH_CAST_METADATA());

        // Adding listeners to THEOplayer basic playback events.
        theoPlayer.addEventListener(PlayerEventTypes.PLAY, event -> Log.i(TAG, "Event: PLAY"));
        theoPlayer.addEventListener(PlayerEventTypes.PLAYING, event -> Log.i(TAG, "Event: PLAYING"));
        theoPlayer.addEventListener(PlayerEventTypes.PAUSE, event -> Log.i(TAG, "Event: PAUSE"));
        theoPlayer.addEventListener(PlayerEventTypes.ENDED, event -> Log.i(TAG, "Event: ENDED"));
        theoPlayer.addEventListener(PlayerEventTypes.ERROR, event -> Log.i(TAG, "Event: ERROR, error=" + event.getErrorObject().getMessage()));
    }

    private void configureChromecast() {
        // Adding listeners to THEOplayer cast events.
        theoChromecast.addEventListener(ChromecastEventTypes.STATECHANGE, event -> Log.i(TAG, "Event: CAST_STATECHANGE, state=" + event.getState()));
        theoChromecast.addEventListener(ChromecastEventTypes.ERROR, event -> Log.i(TAG, "Event: CAST_ERROR, error=" + event.getError()));

        // Some applications that do not require to have MediaRouteButton to control the connection
        // with the Cast Receiver device can use the below APIs instead.
//        theoChromecast.start();
//        theoChromecast.stop();
//        theoChromecast.join();
//        theoChromecast.leave();

        // Some streaming setups requires casting a different stream to a Cast Receiver device
        // than the one playing on a Cast Sender device, e.g. different DRM capabilities.
        // Code below shows how to configure such different stream to cast.
        theoChromecast.setConnectionCallback(new ChromecastConnectionCallback() {
            /**
             * Called after the player has started the connection to the receiver.
             *
             * - At this point we are trying to load the media from the sender to the receiver.
             * - Returning null will behave same as returning the provided SourceDescription.
             *
             * @param sourceDescription The current SourceDescription on the sender device. (<b>Nullable</b>)
             * @return The SourceDescription to be loaded on the receiver device. (<b>Nullable</b>)
             */
            @Nullable
            @Override
            public SourceDescription onStart(@Nullable SourceDescription sourceDescription) {
                return null;
            }

            /**
             * Called after the player has stopped the connection to the receiver.
             *
             * - At this point we are trying to load the media from the receiver to the sender.
             * - Returning null will behave same as returning the provided SourceDescription.
             *
             * @param sourceDescription The current SourceDescription on the receiver device. (<b>Nullable</b>)
             * @return The SourceDescription to be loaded on the sender device. (<b>Nullable</b>)
             */
            @Nullable
            @Override
            public SourceDescription onStop(@Nullable SourceDescription sourceDescription) {
                return null;
            }

            /**
             * Called after the player has joined an already existing connection to the receiver.
             *
             * - At this point it's possible to load a new media from the sender to the receiver.
             * - Returning null will not change the source on the receiver.
             *
             * @param sourceDescription The current SourceDescription on the current sender device. (<b>Nullable</b>)
             * @return The SourceDescription to be loaded on the receiver device. (<b>Nullable</b>)
             */
            @Nullable
            @Override
            public SourceDescription onJoin(@Nullable SourceDescription sourceDescription) {
                return null;
            }

            /**
             * Called after the player has left the connection to the receiver.
             *
             * - At this point we are trying to load the media from the receiver to the sender.
             * - Returning null will behave same as returning the provided SourceDescription.
             *
             * @param sourceDescription The current SourceDescription on the receiver device. (<b>Nullable</b>)
             * @return The SourceDescription to be loaded on the sender device. (<b>Nullable</b>)
             */
            @Nullable
            @Override
            public SourceDescription onLeave(@Nullable SourceDescription sourceDescription) {
                return null;
            }
        });
    }


    // In order to work properly and in sync with the activity lifecycle changes (e.g. device
    // is rotated, new activity is started or app is moved to background) we need to call
    // the "onResume", "onPause" and "onDestroy" methods of the THEOplayerView when the matching
    // activity methods are called.

    @Override
    protected void onPause() {
        super.onPause();
        viewBinding.theoPlayerView.onPause();
    }

    @Override
    protected void onResume() {
        super.onResume();
        viewBinding.theoPlayerView.onResume();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        viewBinding.theoPlayerView.onDestroy();
    }
}

File: google-cast/src/main/java/com/theoplayer/sample/playback/cast/PlayerActivity.kt
package com.theoplayer.sample.playback.cast

import android.content.Context
import android.os.Bundle
import android.util.Log
import androidx.activity.compose.setContent
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Scaffold
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.remember
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.viewinterop.AndroidView
import androidx.fragment.app.FragmentActivity
import androidx.mediarouter.app.MediaRouteButton
import com.google.android.gms.cast.framework.CastButtonFactory
import com.google.android.gms.cast.framework.CastContext
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.theoplayer.android.api.THEOplayerConfig
import com.theoplayer.android.api.THEOplayerGlobal
import com.theoplayer.android.api.THEOplayerView
import com.theoplayer.android.api.cast.CastConfiguration
import com.theoplayer.android.api.cast.CastIntegrationFactory
import com.theoplayer.android.api.cast.CastStrategy
import com.theoplayer.android.api.cast.chromecast.ChromecastConnectionCallback
import com.theoplayer.android.api.event.chromecast.ChromecastEventTypes
import com.theoplayer.android.api.event.player.ErrorEvent
import com.theoplayer.android.api.event.player.PlayerEventTypes
import com.theoplayer.android.api.source.SourceDescription
import com.theoplayer.android.ui.DefaultUI
import com.theoplayer.android.ui.rememberPlayer
import com.theoplayer.android.ui.theme.THEOplayerTheme
import com.theoplayer.sample.common.AppTopBar
import com.theoplayer.sample.common.SourceManager

class PlayerActivity : FragmentActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {

        // Check if Cast API is available on the device (Google Play Services is available, CastContext can be initialized, and device is not an Android TV).
         fun isCastAvailable(context: Context): Boolean {
            when (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)) {
                ConnectionResult.SUCCESS -> {
                    try {
                        CastContext.getSharedInstance(context)
                        return true
                    } catch (_: Exception) {
                        return false
                    }
                }
                else -> return false
            }
        }

        // Initialize Chromecast immediately, for automatic receiver discovery to work correctly.
        if (isCastAvailable(applicationContext)) {
            CastContext.getSharedInstance(this)
        }

        // Enable all debug logs from THEOplayer.
        THEOplayerGlobal.getSharedInstance(this).logger.enableAllTags()

        super.onCreate(savedInstanceState)

        setContent {
            val context = LocalContext.current
            val theoplayerView = remember(context) {
                THEOplayerView(context, THEOplayerConfig.Builder().build()).apply {
                    keepScreenOn = true

                    // Allow background playback on the player to prevent Chromecast receiver from
                    // pausing when the app is backgrounded.
                    settings.setAllowBackgroundPlayback(true)

                    // THEOplayer automatically adds all available integrations to the player.
                    // Alternatively, you can set autoIntegrations(false) on your player configuration
                    // and add them manually.
                    val configuration = CastConfiguration.Builder()
                        .castStrategy(CastStrategy.MANUAL)
                        .build()
                    val castIntegration = CastIntegrationFactory.createCastIntegration(
                        this, configuration
                    )
                    player.addIntegration(castIntegration)
                }
            }

            val player = rememberPlayer(theoplayerView)
            val theoPlayer = theoplayerView.player
            val theoChromecast = theoplayerView.cast.chromecast

            LaunchedEffect(player) {
                // Coupling the orientation of the device with the fullscreen state.
                // The player will go fullscreen when the device is rotated to landscape
                // and will also exit fullscreen when the device is rotated back to portrait.
                theoplayerView.fullScreenManager.isFullScreenOrientationCoupled = true

                // Configuring THEOplayer with defined SourceDescription object.
                theoPlayer.source = SourceManager.BIG_BUCK_BUNNY_HLS_WITH_CAST_METADATA

                // Set autoplay to start video whenever player is visible.
                theoPlayer.isAutoplay = true

                // Attach player event listeners.
                theoPlayer.addEventListener(PlayerEventTypes.SOURCECHANGE) {
                    Log.i(TAG, "Event: SOURCECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.CURRENTSOURCECHANGE) {
                    Log.i(TAG, "Event: CURRENTSOURCECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.LOADEDDATA) {
                    Log.i(TAG, "Event: LOADEDDATA")
                }
                theoPlayer.addEventListener(PlayerEventTypes.LOADEDMETADATA) {
                    Log.i(TAG, "Event: LOADEDMETADATA")
                }
                theoPlayer.addEventListener(PlayerEventTypes.DURATIONCHANGE) {
                    Log.i(TAG, "Event: DURATIONCHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.TIMEUPDATE) {
//                    Log.i(TAG, "Event: TIMEUPDATE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PLAY) {
                    Log.i(TAG, "Event: PLAY")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PLAYING) {
                    Log.i(TAG, "Event: PLAYING")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PAUSE) {
                    Log.i(TAG, "Event: PAUSE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.SEEKING) {
                    Log.i(TAG, "Event: SEEKING")
                }
                theoPlayer.addEventListener(PlayerEventTypes.SEEKED) {
                    Log.i(TAG, "Event: SEEKED")
                }
                theoPlayer.addEventListener(PlayerEventTypes.WAITING) {
                    Log.i(TAG, "Event: WAITING")
                }
                theoPlayer.addEventListener(PlayerEventTypes.READYSTATECHANGE) {
                    Log.i(TAG, "Event: READYSTATECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.PRESENTATIONMODECHANGE) {
                    Log.i(TAG, "Event: PRESENTATIONMODECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.VOLUMECHANGE) {
                    Log.i(TAG, "Event: VOLUMECHANGE")
                }
                theoPlayer.addEventListener(PlayerEventTypes.ENDED) {
                    Log.i(TAG, "Event: ENDED")
                }
                theoPlayer.addEventListener(PlayerEventTypes.ERROR) { event: ErrorEvent ->
                    Log.i(TAG, "Event: ERROR, error=" + event.errorObject)
                }

                // Attach Chromecast event listeners.
                theoChromecast.addEventListener(ChromecastEventTypes.STATECHANGE) {
                    Log.i(TAG, "Event: CAST_STATECHANGE, state=" + it.state)
                }
                theoChromecast.addEventListener(ChromecastEventTypes.ERROR) {
                    Log.i(TAG, "Event: CAST_ERROR, error=" + it.error)
                }

                // Some applications that do not require a MediaRouteButton to control the connection
                // with the Cast Receiver device can use the below APIs instead.
//                theoChromecast.start()
//                theoChromecast.stop()
//                theoChromecast.join()
//                theoChromecast.leave()

                // Some streaming setups require casting a different stream to a Cast Receiver device
                // than the one playing on a Cast Sender device, e.g. different DRM capabilities.
                // Code below shows how to configure such a different stream to cast.
                theoChromecast.setConnectionCallback(object : ChromecastConnectionCallback {
                    override fun onStart(sourceDescription: SourceDescription?): SourceDescription? {
                        return null
                    }

                    override fun onStop(sourceDescription: SourceDescription?): SourceDescription? {
                        return null
                    }

                    override fun onJoin(sourceDescription: SourceDescription?): SourceDescription? {
                        return null
                    }

                    override fun onLeave(sourceDescription: SourceDescription?): SourceDescription? {
                        return null
                    }
                })
            }

            THEOplayerTheme(useDarkTheme = true) {
                Scaffold(
                    topBar = {
                        AppTopBar(actions = {
                            AndroidView(
                                // This is a custom MediaRouterButton that is used to control the
                                // connection with the Cast Receiver device. Open Video UI already
                                // provides a default MediaRouterButton implementation, but you can
                                // also create your own custom button by using the
                                // CastButtonFactory.setUpMediaRouteButton() method.
                                factory = { _ ->
                                    MediaRouteButton(context).apply {
                                        if (isCastAvailable(applicationContext)) {
                                            CastButtonFactory.setUpMediaRouteButton(context, this)
                                        }
                                    }
                                }
                            )
                        })
                    }
                ) { padding ->
                    DefaultUI(
                        modifier = Modifier
                            .padding(padding)
                            .fillMaxSize(),
                        player = player
                    )
                }
            }
        }
    }

    companion object {
        private val TAG: String = PlayerActivity::class.java.simpleName
    }
}
